crates/core/src/node/sealer.rs

```diff
@@ -1,31 +1,115 @@
-use crate::node::pool::{TxBatch, TxPool};
+use super::inner::node_executor::{Command, NodeExecutorHandle};
+use super::pool::{TxBatch, TxPool};
 use futures::channel::mpsc::Receiver;
+use futures::future::BoxFuture;
 use futures::stream::{Fuse, StreamExt};
 use futures::task::AtomicWaker;
 use futures::Stream;
+use std::future::Future;
 use std::pin::Pin;
 use std::sync::{Arc, RwLock};
 use std::task::{Context, Poll};
 use std::time::Duration;
+use tokio::sync::mpsc;
 use tokio::time::{Interval, MissedTickBehavior};
 use zksync_types::H256;
 
-#[derive(Clone, Debug)]
+// TODO: `BlockSealer` is probably a bad name as this doesn't actually seal blocks, just decides
+//       that certain tx batch needs to be sealed. The actual sealing is handled in `NodeExecutor`.
+//       Consider renaming.
+#[pin_project::pin_project]
 pub struct BlockSealer {
-    /// The mode this sealer currently operates in
-    mode: Arc<RwLock<BlockSealerMode>>,
-    /// Used for task wake up when the sealing mode was forcefully changed
-    waker: Arc<AtomicWaker>,
+    /// Block sealer state (externally mutable).
+    state: BlockSealerState,
+    /// Pool where block sealer is sourcing transactions from.
+    pool: TxPool,
+    /// Node handle to be used when a block needs to be sealed.
+    node_handle: NodeExecutorHandle,
+    /// Future that is sending the next seal command to [`super::NodeExecutor`]
+    #[pin]
+    future: Option<BoxFuture<'static, Result<(), mpsc::error::SendError<Command>>>>,
 }
 
 impl BlockSealer {
-    pub fn new(mode: BlockSealerMode) -> Self {
-        Self {
+    pub fn new(
+        mode: BlockSealerMode,
+        pool: TxPool,
+        node_handle: NodeExecutorHandle,
+    ) -> (Self, BlockSealerState) {
+        let state = BlockSealerState {
             mode: Arc::new(RwLock::new(mode)),
             waker: Arc::new(AtomicWaker::new()),
+        };
+        (
+            Self {
+                state: state.clone(),
+                pool,
+                node_handle,
+                future: None,
+            },
+            state,
+        )
+    }
+}
+
+impl Future for BlockSealer {
```

Comment:
Similarly -- cannot we express this as a simple loop?

Reply:
Yeah, same thing here